README.md
# COMS327Spring2019

bibus_parker.assignment-2.0/CHANGELOG
commit fe65b1266e3b932590570cc0562b415fcc8efc09
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sun Apr 14 15:17:33 2019 -0500

    Added README and CHANGELOG to assignment 2.0. Also added help flag info to the usage display.

commit 80a4c0acca83957fa62bf7c5e3a95950cbf64192
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sun Apr 14 14:51:43 2019 -0500

    Moved everything to the the assignment directory.

commit 0923234283cfb4fb3dceccecad5aaf5714cfe167
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sun Apr 14 11:31:16 2019 -0500

    added checks for the speed and damage to create minimums.

commit 0b3c2995921232272af0133cba2fe3262fd8a207
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sun Apr 14 11:18:47 2019 -0500

    Updated CHANGELOG.

commit 9c6866d47dc9e0ed463b3d11b0b17b6d40aa3fa8
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sun Apr 14 11:17:50 2019 -0500

    Added README and empty changelog file.

commit c5ffbf35b11d411c7a102c2fbfa61204fe7d9074
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sun Apr 14 00:28:19 2019 -0500

    Made some changes to more closely match the specs.

commit d5651fec554764f437974f476b206b93e9f5ca48
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 23:52:52 2019 -0500

    Think I got everything where I want it, will finish readme and CHANGELOG tomorrow, might add in a help menu.

commit e7246f6d92d1cc84ceaeb99d241059c96f5ad97a
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 23:31:19 2019 -0500

    Everything seems to more or less be working, going to add damage things at the top of the screen for battle stuff.

commit 4db125a77e834a13785e177e1df149a4cd8c1670
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 23:13:29 2019 -0500

    Figured out a possible issue to why it was crashing.

commit fdfede5c28196a05f91363d73813c776a82b1b61
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 22:39:01 2019 -0500

    Combat seems to be working properly, going to do some more testing to see if the crashes are pyrite or the program.

commit 8a66c26552ab12599f71a7aa712af19c935824f5
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 21:12:06 2019 -0500

    The characters are no longer dying right away and the hp and damage seems to be working.

commit b9192b2cf2f3e9c7b8126b6ba7e67969b9709f66
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 20:59:37 2019 -0500

    Previous commit was partial mistake but am going to leave it as I want the changes regardless.

commit b2d1d25c9984ba8846809c70972f341e9b0fb93a
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 20:59:04 2019 -0500

    Got the NPC's to not kill each other and to displace each other.

commit 417aa0cadafda03eab004878899087c0c7c830a1
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 20:12:10 2019 -0500

    Got the NPC's to not kill each other and to displace each other.

commit 20e4b54d23ac81c20bd089b8d2320136f78ce129
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 19:14:32 2019 -0500

    Got the speed bonuses applied to the PC.

commit e96bf737f710d5782ff5d98748f30f7089c20dbf
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 18:54:25 2019 -0500

    Added speed and damage information to the descriptions.

commit 0b845c3fa1ea41a12b51357867e883ac46b7aab0
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 18:35:46 2019 -0500

    Added message for incorrect spot for L command. Everything works, may go through and change all description things to also print some characteristics.

commit be1059b8116c06c7e174a4ed96d0134fafc1d572
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 18:30:12 2019 -0500

    Got the L command to work properly, going to see if I can get it to display the invalid message.

commit d5c35216c0e1054d1e3a9c8b41e4205018555ef6
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 17:14:53 2019 -0500

    Fixed the instructions to include pressing enter, and some beginning things for the L command.

commit 1e8755127b65395db7d7fd992590ace8822031ce
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 15:44:51 2019 -0500

    Got the unequiping to work properly.

commit 6d7259a1a5fdc5978c2fe5d5923bf29c5753f98f
Author: Parker Bibus <pbibus@hotmail.com>
Date:   Sat Apr 13 15:05:23 2019 -0500

    Got the ring equipping to work! YAY!


bibus_parker.assignment-1.08/io.cpp
#include <unistd.h>
#include <ncurses.h>
#include <ctype.h>
#include <stdlib.h>
#include <iostream>

#include "io.h"
#include "move.h"
#include "path.h"
#include "pc.h"
#include "utils.h"
#include "dungeon.h"
#include "character.h"

/* Same ugly hack we did in path.c */
static dungeon *thedungeon;

typedef struct io_message {
  /* Will print " --more-- " at end of line when another message follows. *
   * Leave 10 extra spaces for that.                                      */
  char msg[71];
  struct io_message *next;
} io_message_t;

static io_message_t *io_head, *io_tail;

void io_init_terminal(void)
{
  initscr();
  raw();
  noecho();
  curs_set(0);
  keypad(stdscr, TRUE);
  start_color();
  
  init_pair(COLOR_RED, COLOR_RED, COLOR_BLACK);
  init_pair(COLOR_GREEN, COLOR_GREEN, COLOR_BLACK);
  init_pair(COLOR_YELLOW, COLOR_YELLOW, COLOR_BLACK);
  init_pair(COLOR_BLUE, COLOR_BLUE, COLOR_BLACK);
  init_pair(COLOR_MAGENTA, COLOR_MAGENTA, COLOR_BLACK);
  init_pair(COLOR_CYAN, COLOR_CYAN, COLOR_BLACK);
  init_pair(COLOR_WHITE, COLOR_WHITE, COLOR_BLACK);
}

void io_reset_terminal(void)
{
  endwin();

  while (io_head) {
    io_tail = io_head;
    io_head = io_head->next;
    free(io_tail);
  }
  io_tail = NULL;
}

void io_queue_message(const char *format, ...)
{
  io_message_t *tmp;
  va_list ap;

  if (!(tmp = (io_message_t *) malloc(sizeof (*tmp)))) {
    perror("malloc");
    exit(1);
  }

  tmp->next = NULL;

  va_start(ap, format);

  vsnprintf(tmp->msg, sizeof (tmp->msg), format, ap);

  va_end(ap);

  if (!io_head) {
    io_head = io_tail = tmp;
  } else {
    io_tail->next = tmp;
    io_tail = tmp;
  }
}

static void io_print_message_queue(uint32_t y, uint32_t x)
{
  while (io_head) {
    io_tail = io_head;
    attron(COLOR_PAIR(COLOR_CYAN));
    mvprintw(y, x, "%-80s", io_head->msg);
    attroff(COLOR_PAIR(COLOR_CYAN));
    io_head = io_head->next;
    if (io_head) {
      attron(COLOR_PAIR(COLOR_CYAN));
      mvprintw(y, x + 70, "%10s", " --more-- ");
      attroff(COLOR_PAIR(COLOR_CYAN));
      refresh();
      getch();
    }
    free(io_tail);
  }
  io_tail = NULL;
}

void io_display_tunnel(dungeon *d)
{
  uint32_t y, x;
  clear();
  for (y = 0; y < DUNGEON_Y; y++) {
    for (x = 0; x < DUNGEON_X; x++) {
      if (charxy(x, y) == d->PC) {
        mvaddch(y + 1, x, charxy(x, y)->symbol);
      } else if (hardnessxy(x, y) == 255) {
        mvaddch(y + 1, x, '*');
      } else {
        mvaddch(y + 1, x, '0' + (d->pc_tunnel[y][x] % 10));
      }
    }
  }
  refresh();
}

void io_display_distance(dungeon *d)
{
  uint32_t y, x;
  clear();
  for (y = 0; y < DUNGEON_Y; y++) {
    for (x = 0; x < DUNGEON_X; x++) {
      if (charxy(x, y)) {
        mvaddch(y + 1, x, charxy(x, y)->symbol);
      } else if (hardnessxy(x, y) != 0) {
        mvaddch(y + 1, x, ' ');
      } else {
        mvaddch(y + 1, x, '0' + (d->pc_distance[y][x] % 10));
      }
    }
  }
  refresh();
}

static char hardness_to_char[] =
  "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

void io_display_hardness(dungeon *d)
{
  uint32_t y, x;
  clear();
  for (y = 0; y < DUNGEON_Y; y++) {
    for (x = 0; x < DUNGEON_X; x++) {
      /* Maximum hardness is 255.  We have 62 values to display it, but *
       * we only want one zero value, so we need to cover [1,255] with  *
       * 61 values, which gives us a divisor of 254 / 61 = 4.164.       *
       * Generally, we want to avoid floating point math, but this is   *
       * not gameplay, so we'll make an exception here to get maximal   *
       * hardness display resolution.                                   */
      mvaddch(y + 1, x, (d->hardness[y][x]                             ?
                         hardness_to_char[1 + (int) ((d->hardness[y][x] /
                                                      4.2))] : ' '));
    }
  }
  refresh();
}

static int compare_monster_distance(const void *v1, const void *v2)
{
  const character *const *c1 = (const character *const *) v1;
  const character *const *c2 = (const character *const *) v2;

  return (thedungeon->pc_distance[(*c1)->position[dim_y]]
                                 [(*c1)->position[dim_x]] -
          thedungeon->pc_distance[(*c2)->position[dim_y]]
                                 [(*c2)->position[dim_x]]);
}

static character *io_nearest_visible_monster(dungeon *d)
{
  character **c, *n;
  uint32_t x, y, count, i;

  c = (character **) malloc(d->num_monsters * sizeof (*c));

  /* Get a linear list of monsters */
  for (count = 0, y = 1; y < DUNGEON_Y - 1; y++) {
    for (x = 1; x < DUNGEON_X - 1; x++) {
      if (d->character_map[y][x] && d->character_map[y][x] != d->PC) {
        c[count++] = d->character_map[y][x];
      }
    }
  }

  /* Sort it by distance from PC */
  thedungeon = d;
  qsort(c, count, sizeof (*c), compare_monster_distance);

  for (n = NULL, i = 0; i < count; i++) {
    if (can_see(d, character_get_pos(d->PC), character_get_pos(c[i]), 1, 0)) {
      n = c[i];
      break;
    }
  }

  free(c);

  return n;
}

void io_display(dungeon *d)
{
  uint32_t y, x;
  uint32_t illuminated;
  character *c;
  int32_t visible_monsters;

  clear();
  for (visible_monsters = -1, y = 0; y < 21; y++) {
    for (x = 0; x < 80; x++) {
      if ((illuminated = is_illuminated(d->PC, y, x))) {
        attron(A_BOLD);
      }
      if (d->character_map[y][x] &&
           can_see(d,
                  character_get_pos(d->PC),
                  character_get_pos(d->character_map[y][x]),
                  1, 0)) {
                 attron(COLOR_PAIR(d->character_map[y][x]->color.at(0)));
       mvaddch(y + 1, x,
                d->character_map[y][x]->symbol);
                attroff(COLOR_PAIR(d->character_map[y][x]->color.at(0)));
        visible_monsters++;
      } else if (d->item_map[y][x] &&
           can_see(d,
                  character_get_pos(d->PC),
                  item_get_pos(d->item_map[y][x]),
                  1, 0)) {
                attron(COLOR_PAIR(d->item_map[y][x]->color));
       mvaddch(y + 1, x,
                  item_get_symbol(d->item_map[y][x]));
                attroff(COLOR_PAIR(d->item_map[y][x]->color));
      } else {
        switch (pc_learned_terrain(d->PC, y, x)) {
        case ter_wall:
        case ter_wall_immutable:
        case ter_unknown:
          mvaddch(y + 1, x, ' ');
          break;
        case ter_floor:
        case ter_floor_room:
          mvaddch(y + 1, x, '.');
          break;
        case ter_floor_hall:
          mvaddch(y + 1, x, '#');
          break;
        case ter_debug:
          mvaddch(y + 1, x, '*');
          break;
        case ter_stairs_up:
          mvaddch(y + 1, x, '<');
          break;
        case ter_stairs_down:
          mvaddch(y + 1, x, '>');
          break;
        default:
 /* Use zero as an error symbol, since it stands out somewhat, and it's *
  * not otherwise used.                                                 */
          mvaddch(y + 1, x, '0');
        }
      }
      if (illuminated) {
        attroff(A_BOLD);
      }
    }
  }

  mvprintw(23, 1, "PC position is (%2d,%2d).",
           d->PC->position[dim_x], d->PC->position[dim_y]);
  mvprintw(22, 1, "%d known %s.", visible_monsters,
           visible_monsters > 1 ? "monsters" : "monster");
  mvprintw(22, 30, "Nearest visible monster: ");
  if ((c = io_nearest_visible_monster(d))) {
    attron(COLOR_PAIR(COLOR_RED));
    mvprintw(22, 55, "%c at %d %c by %d %c.",
             c->symbol,
             abs(c->position[dim_y] - d->PC->position[dim_y]),
             ((c->position[dim_y] - d->PC->position[dim_y]) <= 0 ?
              'N' : 'S'),
             abs(c->position[dim_x] - d->PC->position[dim_x]),
             ((c->position[dim_x] - d->PC->position[dim_x]) <= 0 ?
              'W' : 'E'));
    attroff(COLOR_PAIR(COLOR_RED));
  } else {
    attron(COLOR_PAIR(COLOR_BLUE));
    mvprintw(22, 55, "NONE.");
    attroff(COLOR_PAIR(COLOR_BLUE));
  }
  
  io_print_message_queue(0, 0);

  refresh();
}

void io_display_no_fog(dungeon *d)
{
  uint32_t y, x;
  character *c;

  clear();
  for (y = 0; y < 21; y++) {
    for (x = 0; x < 80; x++) {
      if (d->character_map[y][x]) {
        attron(COLOR_PAIR(d->character_map[y][x]->color.at(0)));
        mvaddch(y + 1, x, d->character_map[y][x]->symbol);
        attroff(COLOR_PAIR(d->character_map[y][x]->color.at(0)));
      } else if (d->item_map[y][x]) {
        attron(COLOR_PAIR(d->item_map[y][x]->color));
        mvaddch(y + 1, x, d->item_map[y][x]->symbol);
        attroff(COLOR_PAIR(d->item_map[y][x]->color));
      } else {
        switch (mapxy(x, y)) {
        case ter_wall:
        case ter_wall_immutable:
          mvaddch(y + 1, x, ' ');
          break;
        case ter_floor:
        case ter_floor_room:
          mvaddch(y + 1, x, '.');
          break;
        case ter_floor_hall:
          mvaddch(y + 1, x, '#');
          break;
        case ter_debug:
          mvaddch(y + 1, x, '*');
          break;
        case ter_stairs_up:
          mvaddch(y + 1, x, '<');
          break;
        case ter_stairs_down:
          mvaddch(y + 1, x, '>');
          break;
        default:
 /* Use zero as an error symbol, since it stands out somewhat, and it's *
  * not otherwise used.                                                 */
          mvaddch(y + 1, x, '0');
        }
      }
    }
  }

  mvprintw(23, 1, "PC position is (%2d,%2d).",
           d->PC->position[dim_x], d->PC->position[dim_y]);
  mvprintw(22, 1, "%d %s.", d->num_monsters,
           d->num_monsters > 1 ? "monsters" : "monster");
  mvprintw(22, 30, "Nearest visible monster: ");
  if ((c = io_nearest_visible_monster(d))) {
    attron(COLOR_PAIR(COLOR_RED));
    mvprintw(22, 55, "%c at %d %c by %d %c.",
             c->symbol,
             abs(c->position[dim_y] - d->PC->position[dim_y]),
             ((c->position[dim_y] - d->PC->position[dim_y]) <= 0 ?
              'N' : 'S'),
             abs(c->position[dim_x] - d->PC->position[dim_x]),
             ((c->position[dim_x] - d->PC->position[dim_x]) <= 0 ?
              'W' : 'E'));
    attroff(COLOR_PAIR(COLOR_RED));
  } else {
    attron(COLOR_PAIR(COLOR_BLUE));
    mvprintw(22, 55, "NONE.");
    attroff(COLOR_PAIR(COLOR_BLUE));
  }
  
  io_print_message_queue(0, 0);
}

void io_display_monster_list(dungeon *d)
{
  mvprintw(11, 33, " HP:    XXXXX ");
  mvprintw(12, 33, " Speed: XXXXX ");
  mvprintw(14, 27, " Hit any key to continue. ");
  refresh();
  getch();
}

uint32_t io_teleport_pc(dungeon *d)
{
  pair_t dest;
  int c;
  int actual;

  pc_reset_visibility(d->PC);
  io_display_no_fog(d);

  mvprintw(0, 0,
           "Choose a location.  'g' or '.' to teleport to; 'r' for random.");

  dest[dim_y] = d->PC->position[dim_y];
  dest[dim_x] = d->PC->position[dim_x];

  mvaddch(dest[dim_y] + 1, dest[dim_x], '*');
  refresh();

  while ((c = getch()) != 'g' && c != '.' && c != 'r') {
    if (charpair(dest)) {
      actual = character_get_symbol(charpair(dest));
      attron(COLOR_PAIR(charpair(dest)->color.at(0)));
      mvaddch(dest[dim_y] + 1, dest[dim_x], actual);
      attroff(COLOR_PAIR(charpair(dest)->color.at(0)));
    }else if (itemxy(dest[dim_x], dest[dim_y])) {
      actual = item_get_symbol(itemxy(dest[dim_x], dest[dim_y]));
      attron(COLOR_PAIR(itemxy(dest[dim_x], dest[dim_y])->color));
      mvaddch(dest[dim_y] + 1, dest[dim_x], actual);
      attroff(COLOR_PAIR(itemxy(dest[dim_x], dest[dim_y])->color));
    } else {
      switch (mappair(dest)) {
      case ter_wall:
      case ter_wall_immutable:
        actual = ' ';
        break;
      case ter_floor:
      case ter_floor_room:
        actual = '.';
        break;
      case ter_floor_hall:
        actual = '#';
        break;
      case ter_debug:
        actual = '*';
        break;
      case ter_stairs_up:
        actual = '<';
        break;
      case ter_stairs_down:
        actual = '>';
        break;
      default:
        break;
      }      
      mvaddch(dest[dim_y] + 1, dest[dim_x], actual);
    }

    //mvaddch(dest[dim_y] + 1, dest[dim_x], actual);

    switch (c) {
    case '7':
    case 'y':
    case KEY_HOME:
      if (dest[dim_y] > 1) {
        dest[dim_y]--;
      }
      if (dest[dim_x] > 1) {
        dest[dim_x]--;
      }
      break;
    case '8':
    case 'k':
    case KEY_UP:
      if (dest[dim_y] > 1) {
        dest[dim_y]--;
      }
      break;
    case '9':
    case 'u':
    case KEY_PPAGE:
      if (dest[dim_y] > 1) {
        dest[dim_y]--;
      }
      if (dest[dim_x] < DUNGEON_X - 2) {
        dest[dim_x]++;
      }
      break;
    case '6':
    case 'l':
    case KEY_RIGHT:
      if (dest[dim_x] < DUNGEON_X - 2) {
        dest[dim_x]++;
      }
      break;
    case '3':
    case 'n':
    case KEY_NPAGE:
      if (dest[dim_y] < DUNGEON_Y - 2) {
        dest[dim_y]++;
      }
      if (dest[dim_x] < DUNGEON_X - 2) {
        dest[dim_x]++;
      }
      break;
    case '2':
    case 'j':
    case KEY_DOWN:
      if (dest[dim_y] < DUNGEON_Y - 2) {
        dest[dim_y]++;
      }
      break;
    case '1':
    case 'b':
    case KEY_END:
      if (dest[dim_y] < DUNGEON_Y - 2) {
        dest[dim_y]++;
      }
      if (dest[dim_x] > 1) {
        dest[dim_x]--;
      }
      break;
    case '4':
    case 'h':
    case KEY_LEFT:
      if (dest[dim_x] > 1) {
        dest[dim_x]--;
      }
      break;
    }

    mvaddch(dest[dim_y] + 1, dest[dim_x], '*');
    refresh();
  }

  if (c == 'r') {
    do {
      dest[dim_x] = rand_range(1, DUNGEON_X - 2);
      dest[dim_y] = rand_range(1, DUNGEON_Y - 2);
    } while (charpair(dest) || mappair(dest) < ter_floor);
  }

  if (charpair(dest) && charpair(dest) != d->PC) {
    io_queue_message("Teleport failed.  Destination occupied.");
  } else {  
    d->character_map[d->PC->position[dim_y]][d->PC->position[dim_x]] = NULL;
    d->character_map[dest[dim_y]][dest[dim_x]] = d->PC;

    d->PC->position[dim_y] = dest[dim_y];
    d->PC->position[dim_x] = dest[dim_x];
  }

  pc_observe_terrain(d->PC, d);
  dijkstra(d);
  dijkstra_tunnel(d);

  io_display(d);

  return 0;
}

/* Adjectives to describe our monsters */
static const char *adjectives[] = {
  "A menacing ",
  "A threatening ",
  "A horrifying ",
  "An intimidating ",
  "An aggressive ",
  "A frightening ",
  "A terrifying ",
  "A terrorizing ",
  "An alarming ",
  "A dangerous ",
  "A glowering ",
  "A glaring ",
  "A scowling ",
  "A chilling ",
  "A scary ",
  "A creepy ",
  "An eerie ",
  "A spooky ",
  "A slobbering ",
  "A drooling ",
  "A horrendous ",
  "An unnerving ",
  "A cute little ",  /* Even though they're trying to kill you, */
  "A teeny-weenie ", /* they can still be cute!                 */
  "A fuzzy ",
  "A fluffy white ",
  "A kawaii ",       /* For our otaku */
  "Hao ke ai de ",   /* And for our Chinese */
  "Eine liebliche "  /* For our Deutch */
  /* And there's one special case (see below) */
};

static void io_scroll_monster_list(char (*s)[40], uint32_t count)
{
  uint32_t offset;
  uint32_t i;

  offset = 0;

  while (1) {
    for (i = 0; i < 13; i++) {
      mvprintw(i + 6, 19, " %-40s ", s[i + offset]);
    }
    switch (getch()) {
    case KEY_UP:
      if (offset) {
        offset--;
      }
      break;
    case KEY_DOWN:
      if (offset < (count - 13)) {
        offset++;
      }
      break;
    case 27:
      return;
    }

  }
}

static void io_list_monsters_display(dungeon *d,
                                     character **c,
                                     uint32_t count)
{
  uint32_t i;
  char (*s)[40]; /* pointer to array of 40 char */

  s = (char (*)[40]) malloc(count * sizeof (*s));

  mvprintw(3, 19, " %-40s ", "");
  /* Borrow the first element of our array for this string: */
  snprintf(s[0], 40, "You know of %d monsters:", count);
  mvprintw(4, 19, " %-40s ", s);
  mvprintw(5, 19, " %-40s ", "");

  for (i = 0; i < count; i++) {
    snprintf(s[i], 40, "%16s%c: %2d %s by %2d %s",
             (c[i]->symbol == 'd' ? "A tenacious " :
              adjectives[rand() % (sizeof (adjectives) /
                                   sizeof (adjectives[0]))]),
             c[i]->symbol,
             abs(c[i]->position[dim_y] - d->PC->position[dim_y]),
             ((c[i]->position[dim_y] - d->PC->position[dim_y]) <= 0 ?
              "North" : "South"),
             abs(c[i]->position[dim_x] - d->PC->position[dim_x]),
             ((c[i]->position[dim_x] - d->PC->position[dim_x]) <= 0 ?
              "West" : "East"));
    if (count <= 13) {
      /* Handle the non-scrolling case right here. *
       * Scrolling in another function.            */
      mvprintw(i + 6, 19, " %-40s ", s[i]);
    }
  }

  if (count <= 13) {
    mvprintw(count + 6, 19, " %-40s ", "");
    mvprintw(count + 7, 19, " %-40s ", "Hit escape to continue.");
    while (getch() != 27 /* escape */)
      ;
  } else {
    mvprintw(19, 19, " %-40s ", "");
    mvprintw(20, 19, " %-40s ",
             "Arrows to scroll, escape to continue.");
    io_scroll_monster_list(s, count);
  }

  free(s);
}

static void io_list_monsters(dungeon *d)
{
  character **c;
  uint32_t x, y, count;

  c = (character **) malloc(d->num_monsters * sizeof (*c));

  /* Get a linear list of monsters */
  for (count = 0, y = 1; y < DUNGEON_Y - 1; y++) {
    for (x = 1; x < DUNGEON_X - 1; x++) {
      if (d->character_map[y][x] && d->character_map[y][x] != d->PC &&
          can_see(d, character_get_pos(d->PC),
                  character_get_pos(d->character_map[y][x]), 1, 0)) {
        c[count++] = d->character_map[y][x];
      }
    }
  }

  /* Sort it by distance from PC */
  thedungeon = d;
  qsort(c, count, sizeof (*c), compare_monster_distance);

  /* Display it */
  io_list_monsters_display(d, c, count);
  free(c);

  /* And redraw the dungeon */
  io_display(d);
}

void io_handle_input(dungeon *d)
{
  uint32_t fail_code;
  int key;

  do {
    switch (key = getch()) {
    case '7':
    case 'y':
    case KEY_HOME:
      fail_code = move_pc(d, 7);
      break;
    case '8':
    case 'k':
    case KEY_UP:
      fail_code = move_pc(d, 8);
      break;
    case '9':
    case 'u':
    case KEY_PPAGE:
      fail_code = move_pc(d, 9);
      break;
    case '6':
    case 'l':
    case KEY_RIGHT:
      fail_code = move_pc(d, 6);
      break;
    case '3':
    case 'n':
    case KEY_NPAGE:
      fail_code = move_pc(d, 3);
      break;
    case '2':
    case 'j':
    case KEY_DOWN:
      fail_code = move_pc(d, 2);
      break;
    case '1':
    case 'b':
    case KEY_END:
      fail_code = move_pc(d, 1);
      break;
    case '4':
    case 'h':
    case KEY_LEFT:
      fail_code = move_pc(d, 4);
      break;
    case '5':
    case ' ':
    case '.':
    case KEY_B2:
      fail_code = 0;
      break;
    case '>':
      fail_code = move_pc(d, '>');
      break;
    case '<':
      fail_code = move_pc(d, '<');
      break;
    case 'Q':
      d->quit = 1;
      fail_code = 0;
      break;
    case 'T':
      /* New command.  Display the distances for tunnelers.             */
      io_display_tunnel(d);
      fail_code = 1;
      break;
    case 'D':
      /* New command.  Display the distances for non-tunnelers.         */
      io_display_distance(d);
      fail_code = 1;
      break;
    case 'H':
      /* New command.  Display the hardnesses.                          */
      io_display_hardness(d);
      fail_code = 1;
      break;
    case 's':
      /* New command.  Return to normal display after displaying some   *
       * special screen.                                                */
      io_display(d);
      fail_code = 1;
      break;
    case 'L':
      fail_code = 1;
      break;
    case 'g':
      /* Teleport the PC to a random place in the dungeon.              */
      io_teleport_pc(d);
      fail_code = 1;
      break;
    case 'f':
      io_display_no_fog(d);
      fail_code = 1;
      break;
     case 'm':
      io_list_monsters(d);
      fail_code = 1;
      break;
    case 'q':
      /* Demonstrate use of the message queue.  You can use this for *
       * printf()-style debugging (though gdb is probably a better   *
       * option.  Not that it matterrs, but using this command will  *
       * waste a turn.  Set fail_code to 1 and you should be able to *
       * figure out why I did it that way.                           */
      io_queue_message("This is the first message.");
      io_queue_message("Since there are multiple messages, "
                       "you will see \"more\" prompts.");
      io_queue_message("You can use any key to advance through messages.");
      io_queue_message("Normal gameplay will not resume until the queue "
                       "is empty.");
      io_queue_message("Long lines will be truncated, not wrapped.");
      io_queue_message("io_queue_message() is variadic and handles "
                       "all printf() conversion specifiers.");
      io_queue_message("Did you see %s?", "what I did there");
      io_queue_message("When the last message is displayed, there will "
                       "be no \"more\" prompt.");
      io_queue_message("Have fun!  And happy printing!");
      fail_code = 0;
      break;
    default:
      /* Also not in the spec.  It's not always easy to figure out what *
       * key code corresponds with a given keystroke.  Print out any    *
       * unhandled key here.  Not only does it give a visual error      *
       * indicator, but it also gives an integer value that can be used *
       * for that key in this (or other) switch statements.  Printed in *
       * octal, with the leading zero, because ncurses.h lists codes in *
       * octal, thus allowing us to do reverse lookups.  If a key has a *
       * name defined in the header, you can use the name here, else    *
       * you can directly use the octal value.                          */
      mvprintw(0, 0, "Unbound key: %#o ", key);
      fail_code = 1;
    }
  } while (fail_code);
}
